python/pyspark/ml/functions.py

```diff
@@ -106,6 +117,602 @@ def array_to_vector(col: Column) -> Column:
+        takes one or more numpy arrays as input and returns one of the following:
+        - a numpy array (for a single output)
+        - a dictionary of named numpy arrays (for multiple outputs)
+        - a row-oriented list of dictionaries (for multiple outputs).
+        For a dictionary of named numpy arrays, the arrays can only be one or two dimensional, since
+        higher dimension arrays are not supported.  For a row-oriented list of dictionaries, each
+        element in the dictionary must be either a scalar or one-dimensional array.
+    return_type : :class:`pspark.sql.types.DataType` or str.
+        Spark SQL datatype for the expected output:
+        - Scalar (e.g. IntegerType, FloatType) --> 1-dim numpy array.
+        - ArrayType --> 2-dim numpy array.
+        - StructType --> dict with keys matching struct fields.
+        - StructType --> list of dict with keys matching struct fields, for models like the
+        [Huggingface pipeline for sentiment analysis](https://huggingface.co/docs/transformers/quicktour#pipeline-usage]  # noqa: E501
+    batch_size : int
+        Batch size to use for inference, note that this is typically a limitation of the model
+        and/or the hardware resources and is usually smaller than the Spark partition size.
+    input_tensor_shapes: List[List[int] | None] | Mapping[int, List[int]] | None
+        Optional input tensor shapes for models with tensor inputs.  This can be a list of shapes,
+        where each shape is a list of integers or None (for scalar inputs).  Alternatively, this
+        can be represented by a "sparse" dictionary, where the keys are the integer indices of the
+        inputs, and the values are the shapes.  Each tensor input value in the Spark DataFrame must
+        be represented as a single column containing a flattened 1-D array.  The provided
+        input_tensor_shapes will be used to reshape the flattened array into expected tensor shape.
+        For the list form, the order of the tensor shapes must match the order of the selected
+        DataFrame columns.  The batch dimension (typically -1 or None in the first dimension) should
+        not be included, since it will be determined by the batch_size argument.  Tabular datasets
+        with scalar-valued columns should not provide this argument.
+
+    Returns
+    -------
+    A pandas_udf for predicting a batch.
+    """
+    # generate a new uuid each time this is invoked on the driver to invalidate executor-side cache.
+    model_uuid = uuid.uuid4()
+
+    def predict(data: Iterator[Union[pd.Series, pd.DataFrame]]) -> Iterator[pd.DataFrame]:
+        # TODO: adjust return type hint when Iterator[Union[pd.Series, pd.DataFrame]] is supported
+        from pyspark.ml.model_cache import ModelCache
+
+        # get predict function (from cache or from running user-provided predict_batch_fn)
+        predict_fn = ModelCache.get(model_uuid)
```

Comment:
I synced with @mengxr offline,
@mengxr suggests to remove the ModelCache, because:
* It will cause memory leak if spark python worker reuse = true.
* Without it, the scalar iterator UDF still runs the entire partition. Users can merge partitions to reuse the loaded model.

Reply:
I went ahead and converted the ModelCache into an LRU cache with a limit of 8 items.  I think this should be still be useful for end users while avoiding any memory leaks due to an unbounded cache.  And worst case, users can still merge partitions as you mentioned.